Mathlib/Data/Nat/Factorization/PrimePow.lean

```diff
@@ -0,0 +1,158 @@
+    IsPrimePow n ↔ ∃ p : ℕ, p.Prime ∧ ord_compl[p] n = 1 := by
+  refine' ⟨fun h => IsPrimePow.exists_ord_compl_eq_one h, fun h => _⟩
+  rcases h with ⟨p, pp, h⟩
+  rw [isPrimePow_nat_iff]
+  rw [← Nat.eq_of_dvd_of_div_eq_one (Nat.ord_proj_dvd n p) h] at hn⊢
+  refine' ⟨p, n.factorization p, pp, _, by simp⟩
+  contrapose! hn
+  simp [le_zero_iff.1 hn]
+#align exists_ord_compl_eq_one_iff_is_prime_pow exists_ord_compl_eq_one_iff_isPrimePow
+
+/-- An equivalent definition for prime powers: `n` is a prime power iff there is a unique prime
+dividing it. -/
+theorem isPrimePow_iff_unique_prime_dvd {n : ℕ} : IsPrimePow n ↔ ∃! p : ℕ, p.Prime ∧ p ∣ n := by
+  rw [isPrimePow_nat_iff]
+  constructor
+  · rintro ⟨p, k, hp, hk, rfl⟩
+    refine' ⟨p, ⟨hp, dvd_pow_self _ hk.ne'⟩, _⟩
+    rintro q ⟨hq, hq'⟩
+    exact (Nat.prime_dvd_prime_iff_eq hq hp).1 (hq.dvd_of_dvd_pow hq')
+  rintro ⟨p, ⟨hp, hn⟩, hq⟩
+  rcases eq_or_ne n 0 with (rfl | hn₀)
+  · cases (hq 2 ⟨Nat.prime_two, dvd_zero 2⟩).trans (hq 3 ⟨Nat.prime_three, dvd_zero 3⟩).symm
+  refine' ⟨p, n.factorization p, hp, hp.factorization_pos_of_dvd hn₀ hn, _⟩
+  simp only [and_imp] at hq
+  apply Nat.dvd_antisymm (Nat.ord_proj_dvd _ _)
+  -- We need to show n ∣ p ^ n.factorization p
+  apply Nat.dvd_of_factors_subperm hn₀
+  rw [hp.factors_pow, List.subperm_ext_iff]
+  intro q hq'
+  rw [Nat.mem_factors hn₀] at hq'
+  cases hq _ hq'.1 hq'.2
+  simp
+#align is_prime_pow_iff_unique_prime_dvd isPrimePow_iff_unique_prime_dvd
+
+theorem isPrimePow_pow_iff {n k : ℕ} (hk : k ≠ 0) : IsPrimePow (n ^ k) ↔ IsPrimePow n := by
+  simp only [isPrimePow_iff_unique_prime_dvd]
+  apply exists_unique_congr
+  simp only [and_congr_right_iff]
+  intro p hp
+  exact ⟨hp.dvd_of_dvd_pow, fun t => t.trans (dvd_pow_self _ hk)⟩
+#align is_prime_pow_pow_iff isPrimePow_pow_iff
+
+theorem Nat.coprime.isPrimePow_dvd_mul {n a b : ℕ} (hab : Nat.coprime a b) (hn : IsPrimePow n) :
+    n ∣ a * b ↔ n ∣ a ∨ n ∣ b := by
+  rcases eq_or_ne a 0 with (rfl | ha)
+  · simp only [Nat.coprime_zero_left] at hab
+    simp [hab, Finset.filter_singleton, not_isPrimePow_one]
+  rcases eq_or_ne b 0 with (rfl | hb)
+  · simp only [Nat.coprime_zero_right] at hab
+    simp [hab, Finset.filter_singleton, not_isPrimePow_one]
+  refine'
+    ⟨_, fun h =>
+      Or.elim h (fun i => i.trans ((@dvd_mul_right a b a hab).mpr (dvd_refl a)))
+          fun i => i.trans ((@dvd_mul_left a b b hab.symm).mpr (dvd_refl b))⟩
+  obtain ⟨p, k, hp, hk, rfl⟩ := (isPrimePow_nat_iff _).1 hn
+  simp only [hp.pow_dvd_iff_le_factorization (mul_ne_zero ha hb), Nat.factorization_mul ha hb,
+    hp.pow_dvd_iff_le_factorization ha, hp.pow_dvd_iff_le_factorization hb, Pi.add_apply,
+    Finsupp.coe_add]
+  have : a.factorization p = 0 ∨ b.factorization p = 0 := by
+    rw [← Finsupp.not_mem_support_iff, ← Finsupp.not_mem_support_iff, ← not_and_or, ←
+      Finset.mem_inter]
+    exact fun t => (Nat.factorization_disjoint_of_coprime hab).le_bot t
```

Comment:
```suggestion
    intro t
    -- porting note: used to be `exact`, but the definition of `∈` has changed.
    simpa using (Nat.factorization_disjoint_of_coprime hab).le_bot t
```
How I found this: I checked whether `simp` could do it, and I looked at the lemmas it used:
```
    intro t
    have := (Nat.factorization_disjoint_of_coprime hab).le_bot t
    simp? at this
```
and then I was satisfied that I could use `simpa using` instead, which is more-or-less `simp`-then-`exact`.

This would work too, though I think it's somewhat odd style.
```suggestion
    simpa using fun t => (Nat.factorization_disjoint_of_coprime hab).le_bot t
```

Reply:
Right, I'm a bit dumb, eh?

Reply:
Just haven't seen all of Kyle's tricks yet. :-)